---
title: "RNAseq analysis on mason data"
author: "Jakob Jung"
date: "May 20, 2021"
output:
  pdf_document: default
  html_document:
    df_print: paged
---
Here I perform the downstream analysis of the RNAseq experiment, in which Salmonella 1344 was challenged with PNA targeting the different genes. RNAseq was performed to determine transcriptome changes upon exposure to PNA. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = TRUE)
```

# Packages
I import the packages needed for all analysis. They can all be installed from Bioconductor or CRAN if not statet otherwise:
```{r, message=FALSE}
library(edgeR)
library(circlize)
library(dplyr)
library(ggplot2)
library('RUVSeq')
library(RColorBrewer)
library(oligo)
library(EDASeq)
library(gplots)
library(ggrepel)
library(viridis)
library(svglite)
library(ComplexHeatmap)
library(eulerr)
library(tidyverse)
library(grid)
library(circlize)
```


# Data Acquisition  

I generated a tab file containing gene counts after upstream processing. Upstream processing included folowing steps (starting with fastq-files): 
 
 - BBtools for filtering, trimming and mapping 
 - featureCounts for generating a count matrix


Firstly, I import the gene-wise counts:
```{r}
GenewiseCounts <- read.delim(
  "../data/rna_align/counttable.txt",sep = "\t",
  row.names = 1, header = T, comment.char = "#")

dim(GenewiseCounts)
head(GenewiseCounts[,1:6])
```
\hfill\break

I have to change column names, since they include the whole path:
```{r}
gwc <- GenewiseCounts[,5:length(GenewiseCounts[1,])]
pnapat <- "\\.\\.\\.data\\.rna_align\\.hash(.*)\\.fq\\.gz\\.bam"
colnames (gwc) <- gsub(pnapat,"\\1", colnames(gwc))


gwc_lex <- gwc
```

# Gene length vs. expression:
make plot visualizing length:
```{r}
# normalized (TPM)
gwc_lex <- gwc_lex[, grepl("(control)|(Length)", colnames(gwc_lex))]
lex <- data.frame(length=gwc_lex$Length, counts=rowMeans(gwc_lex[,-1]))

#vert <- data.frame(length=gwc_old_vertis$Length, counts=rowMeans(gwc_old_vertis[,-1]))

gwcnorm_length <- data.frame(sapply(gwc_lex[,-1], function(x) x / (gwc[,1]/1000)))
gwc_tpm <- data.frame(Length = gwc_lex$Length, 
                      sapply(gwcnorm_length, function(x) x * 1e6 / sum(x)), row.names = rownames(gwc_lex))
gwc_tpm[, grepl("(control)|(Length)", colnames(gwc_tpm))]
tpm <- data.frame(length=gwc_tpm$Length, counts=rowMeans(gwc_tpm[,-1]))
```

Make plots:
```{r}
lex %>% ggplot(aes(x=length, y=log10(counts+1))) + geom_point() + 
  scale_x_continuous(limits = c(0,500)) + scale_y_continuous(limits = c(-0.5,5)) + theme_minimal() +
  geom_text_repel(aes( label=ifelse(length<80, rownames(lex), "")), size=2.5, max.overlaps = 15)

tpm %>% ggplot(aes(x=length, y=log10(counts+1))) + geom_point() + 
  scale_x_continuous(limits = c(0,500)) + scale_y_continuous(limits = c(0,5)) + theme_minimal()

#vert %>% ggplot(aes(x=length, y=log10(counts+1))) + geom_point() + 
#  scale_x_continuous(limits = c(0,500)) + scale_y_continuous(limits = c(-0.5,5)) + theme_minimal() +
#  geom_text_repel(aes( label=ifelse(length<80, rownames(lex), "")), size=2.5, max.overlaps = 15)
```


I also create a facor variable for groups of the sample data manually (from assigning sample codes to condition):
```{r}
namess <- read.csv("../data/names.csv")
namess$sample <- gsub("#(.*) (.*)", "\\1_\\2", namess$sample)
namess$sample <- gsub(" ", "", namess$sample)
test_o <- colnames(gwc)[-1]
test_n <- sapply(test_o, function(x) ifelse(x %in% namess$sample,namess[namess$sample==x,]$test, "control" ))
names(test_n) <- NULL
test <- as.factor(gsub("-", "_",test_n))
test
```

As we have 2 different experiments (one with changes in acpP PNA and one with non-essential genes), we will create 2 edgeR objects to analyse the data separateely:
```{r}
acpP_mut_test <- as.factor(as.character(test[grepl("acpP|control", test)]))
non_ess_test <- as.factor(as.character(test[grepl("KFF|control", test)]))

acpP_mut_gwc <- gwc[,grepl("Length|acpP|control", c("Length", as.character(test)))]
non_ess_gwc <- gwc[,grepl("Length|KFF|control", c("Length", as.character(test)))]
```



# DE analysis *acpP* mismatch-PNAs

Now that I have the read count dataframe with sample names, I import them into the edgeR environment:
```{r}
y <- DGEList(acpP_mut_gwc[,-1], group = acpP_mut_test, genes = acpP_mut_gwc[,1,drop=FALSE])
options(digits = 3)
head(y$samples)
```


## Filtering
Now I want to filter out Genes which have very low counts across all libraries. 
I do this by creating a cutoff $$\frac {10} {L} $$
where L is the minimum library size in millions. We delete genes that are below the cutoff in at least 2 libraries:
```{r}
L <- min(y$samples$lib.size) / 1000000
cutoff <- 10/L
keep <- rowSums(cpm(y) > cutoff) >= 5
table(keep)
```
I retain only the unfiltered genes,and delete 519 genes below the threshold:
```{r}
y <- y[keep, , keep.lib.sizes=FALSE]
```



# Design matrix
I create a design matrix for the samples:
```{r}
batches = as.factor(gsub("(\\d_\\d)_.*", "\\1", colnames(y)))

design <- model.matrix(~0+acpP_mut_test+batches)
colnames(design) <- c(levels(acpP_mut_test), gsub("(.*)", "batch_\\1", levels(batches)[-1]))
rownames(design) <- colnames(y$counts)
design[1:5,]
```


# Normalization

I check how the standard TMM normalization of edgeR performs. I start with calculating normalization factors:
```{r, error=FALSE}
y <- calcNormFactors(y)
y <- estimateDisp(y, design, robust = T)
```

And now I create PCA and RLE plots:
```{r}
library(RColorBrewer)
nb.cols <- 18
mycolors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)

plotPCA(cpm(y), col=mycolors[acpP_mut_test])
plotRLE(cpm(y), outline=FALSE, ylim=c(-1, 1), col=mycolors[acpP_mut_test],
        main="RLE", las=2)

# save em as svgs
svg("../analysis/PCA_allsamples.svg")
plotPCA(cpm(y), col=mycolors[acpP_mut_test])
dev.off()
svg("../analysis/RLE_allsamples.svg")
plotRLE(cpm(y), outline=FALSE, ylim=c(-1, 1), col=mycolors[acpP_mut_test],
        main="RLE", las=2)
dev.off()
```

You can see that the TMM was succesful (TMM centers the RLE around 0). The PCA plot shows a slight batch effect, which I address using 


## Differential Expression analysis

Next, I perform differential expression analysis using TMM-normalized dataset:
```{r}
# make a contrast:
con_acpP_mutants <- makeContrasts(acpP_vs_ctrl = acpP - control,
                     acpP_scrambled_vs_ctrl = acpP_scrambled - control,
                     acpP_PNA_mut_1_vs_ctrl = acpP_PNA_mut_1 - control,
                     acpP_PNA_mut_2_vs_ctrl = acpP_PNA_mut_2 - control,
                     acpP_PNA_mut_3_vs_ctrl = acpP_PNA_mut_3 - control,
                     acpP_PNA_mut_4_vs_ctrl = acpP_PNA_mut_4 - control,
                     acpP_PNA_mut_5_vs_ctrl = acpP_PNA_mut_5 - control,
                     acpP_PNA_mut_6_vs_ctrl = acpP_PNA_mut_6 - control,
                     acpP_PNA_mut_7_vs_ctrl = acpP_PNA_mut_7 - control,
                     acpP_PNA_mut_8_vs_ctrl = acpP_PNA_mut_8 - control,
                     acpP_PNA_mut_9_vs_ctrl = acpP_PNA_mut_9 - control,
                     levels = design)


fit <- glmQLFit(y, design, robust = TRUE)


res_acpP_mutants <- list(acpP_vs_ctrl = glmQLFTest(fit, contrast = con_acpP_mutants[,1]),
                 acpP_scrambled_vs_ctrl = glmQLFTest(fit, contrast = con_acpP_mutants[,2]),
                 acpP_PNA_mut_1_vs_ctrl = glmQLFTest(fit, contrast = con_acpP_mutants[,3]), 
                 acpP_PNA_mut_2_vs_ctrl = glmQLFTest(fit, contrast = con_acpP_mutants[,4]), 
                 acpP_PNA_mut_3_vs_ctrl = glmQLFTest(fit, contrast = con_acpP_mutants[,5]), 
                 acpP_PNA_mut_4_vs_ctrl = glmQLFTest(fit, contrast = con_acpP_mutants[,6]),
                 acpP_PNA_mut_5_vs_ctrl = glmQLFTest(fit, contrast = con_acpP_mutants[,7]),
                 acpP_PNA_mut_6_vs_ctrl = glmQLFTest(fit, contrast = con_acpP_mutants[,8]),
                 acpP_PNA_mut_7_vs_ctrl = glmQLFTest(fit, contrast = con_acpP_mutants[,9]),
                 acpP_PNA_mut_8_vs_ctrl = glmQLFTest(fit, contrast = con_acpP_mutants[,10]),
                 acpP_PNA_mut_9_vs_ctrl = glmQLFTest(fit, contrast = con_acpP_mutants[,11])
                )

con_acpP_mutants_scr <- makeContrasts(acpP_vs_scr = acpP - acpP_scrambled,
                     acpP_scrambled_vs_scr = acpP_scrambled - acpP_scrambled,
                     acpP_PNA_mut_1_vs_scr = acpP_PNA_mut_1 - acpP_scrambled,
                     acpP_PNA_mut_2_vs_scr = acpP_PNA_mut_2 - acpP_scrambled,
                     acpP_PNA_mut_3_vs_scr = acpP_PNA_mut_3 - acpP_scrambled,
                     acpP_PNA_mut_4_vs_scr = acpP_PNA_mut_4 - acpP_scrambled,
                     acpP_PNA_mut_5_vs_scr = acpP_PNA_mut_5 - acpP_scrambled,
                     acpP_PNA_mut_6_vs_scr = acpP_PNA_mut_6 - acpP_scrambled,
                     acpP_PNA_mut_7_vs_scr = acpP_PNA_mut_7 - acpP_scrambled,
                     acpP_PNA_mut_8_vs_scr = acpP_PNA_mut_8 - acpP_scrambled,
                     acpP_PNA_mut_9_vs_scr = acpP_PNA_mut_9 - acpP_scrambled,
                     levels = design)

res_acpP_mutants_vs_scr <- list(acpP_vs_scr = glmQLFTest(fit, contrast = con_acpP_mutants_scr[,1]),
                 acpP_PNA_mut_1_vs_scr = glmQLFTest(fit, contrast = con_acpP_mutants_scr[,3]), 
                 acpP_PNA_mut_2_vs_scr = glmQLFTest(fit, contrast = con_acpP_mutants_scr[,4]), 
                 acpP_PNA_mut_3_vs_scr = glmQLFTest(fit, contrast = con_acpP_mutants_scr[,5]), 
                 acpP_PNA_mut_4_vs_scr = glmQLFTest(fit, contrast = con_acpP_mutants_scr[,6]),
                 acpP_PNA_mut_5_vs_scr = glmQLFTest(fit, contrast = con_acpP_mutants_scr[,7]),
                 acpP_PNA_mut_6_vs_scr = glmQLFTest(fit, contrast = con_acpP_mutants_scr[,8]),
                 acpP_PNA_mut_7_vs_scr = glmQLFTest(fit, contrast = con_acpP_mutants_scr[,9]),
                 acpP_PNA_mut_8_vs_scr = glmQLFTest(fit, contrast = con_acpP_mutants_scr[,10]),
                 acpP_PNA_mut_9_vs_scr = glmQLFTest(fit, contrast = con_acpP_mutants_scr[,11])
                )
```





We now create MD, BCV and QLDisp plots to access qualiy of data:
```{r}
plotMD(y, main = "MD-plot")
abline(h=0, col="red", lty=2, lwd=2)
plotBCV(y)
plotQLDisp(fit)
```
The quality looks decent.


Now I create a function which makes nice volcano-plots and run it on all the results (all PNA-samples are compared to water control for DE):
```{r}
do_volcano <- function(restab, targetgene = NULL, pointsize = 2, x_limit = F,y_limit=F, show_sig = F, alpha=0.05, 
                       minlogfc=1, title = "Volcano", off_target_list = NULL, phopq = NULL) {
  cols = c("targets"="red","0 mm"="cyan", "1 mm"="blue","2 mm"="darkblue", "PhoPQ"="darkred")
  rownames(restab) <- gsub("^([^A-Z].+)" , "italic('\\1')" , rownames(restab))
  g_labels <- gsub("^([^A-Z].+)" , "italic('\\1')" , targetgene)
  g = ggplot(restab) +
  geom_point(
    data = restab,
    aes(x = logFC, y = -log10(FDR)),
    color = "darkgrey",
    cex = pointsize
  ) + theme_bw()+ # change theme to standard black&wite.
  geom_hline(yintercept = -log10(alpha),
             color = "black", linetype = 3) +
  geom_vline(xintercept = c(-minlogfc,minlogfc),
             color = "black", linetype = 3) +
  theme(axis.title.x = element_text(size=20),
        axis.title.y = element_text(size=20),
        axis.text = element_text(size=15, colour = "black"),
        panel.background = element_rect(colour = "black"),
        axis.line = element_line(colour = "black"),
        panel.grid.minor.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.major.x =  element_blank(),#element_line(colour="lightgrey", size=0.3),
        panel.grid.major.y = element_blank(),#element_line(colour="lightgrey", size=0.3),
        plot.title = element_text(hjust = 0.5, size = 23))+
  ggtitle(title)+
  xlab(expression("log"[2]*" fold change")) +
  ylab(expression("- log"[10]*" P-value (FDR)"))+
  scale_x_continuous(expand = c(0,0),breaks = seq(-6,6,2), limits = c(-x_limit,x_limit)) +
  scale_y_continuous(expand = c(0, 0),breaks = seq(0,26,2), limits = c(0,y_limit)) 
  
  if (!is.null(phopq)) {
    g <- g + geom_point(
        data = restab[restab$genes %in% phopq,],
        aes(x = logFC, y = -log10(FDR), bg = "PhoPQ"),
        cex = pointsize, pch=21)
  }  
  
  if (is.null(off_target_list)) {
    g <- g + 
      geom_point(
        data = restab[restab$FDR<alpha & restab$logFC < -minlogfc,],
        aes(x = logFC, y = -log10(FDR)),
        color = "blue", 
        cex = pointsize) +
      geom_point(
        data = restab[restab$FDR<alpha & restab$logFC > minlogfc,],
        aes(x = logFC, y = -log10(FDR)),
        color = "red", 
        cex = pointsize) 
  } else{
    # different colors for different mismatches:
    mm321 <- restab[gsub("^([^A-Z].+)" , "italic('\\1')" , unique(unlist(off_target_list[3]))),]
    mm321 <- mm321[mm321$logFC < -1 & mm321$FDR < 0.001,]
    
    g <- g +
      geom_point(
        data = restab[gsub("^([^A-Z].+)" , "italic('\\1')" , unlist(off_target_list[3])),],
        aes(x = logFC, y = -log10(FDR), color = "2 mm"),
        cex = pointsize) +
      geom_point(
        data = restab[gsub("^([^A-Z].+)" , "italic('\\1')" , unlist(off_target_list[2])),],
        aes(x = logFC, y = -log10(FDR), color = "1 mm"),
        cex = pointsize)  +
    #zero mismatches:
      if (length(off_target_list[[1]]) > 0) {
      geom_point(
        data = restab[gsub("^([^A-Z].+)" , "italic('\\1')" , unlist(off_target_list[1])),],
        aes(x = logFC, y = -log10(FDR), color = "0 mm"),
        cex = pointsize) 
      }
    g_labels <- c(g_labels, gsub("^([^A-Z].+)" , "italic('\\1')" , unlist(off_target_list[1])))
    g_labels <- c(g_labels, rownames(mm321))
    g_labels <- g_labels[!grepl("NA", g_labels)]
  }
  
  # show the sign. genes:
  # show the sigficantest genes:
  if(show_sig){
    range01 <- function(x){(x-min(x))/(max(x)-min(x))}
    top_up <- restab[ which(restab$FDR < alpha & restab$logFC > minlogfc),]
    top_down <- restab[ which(restab$FDR < alpha & restab$logFC < -(minlogfc)),]
    
    if (length(rownames(top_up)) > 0 && (length(rownames(top_up)) > 3)){
    logFC.scaled <- range01(top_up$logFC)
    FDR.scaled <- range01(-log(top_up$FDR))
    summ <- (logFC.scaled + FDR.scaled)
    top_up <- top_up[order(-summ),][1:3,]
    }

    if (length(rownames(top_down))>0 && (length(rownames(top_down))> 3)){
      logFC.scaled <- range01(-top_down$logFC)
      FDR.scaled <- range01(-log(top_down$FDR))
      summ <- (logFC.scaled + FDR.scaled)
      top_down <- top_down[order(-summ),][1:3,]
    }

    top_peaks <- rbind(top_up, top_down)
    top_peaks <- na.omit(top_peaks)


    g_labels <- c(g_labels, rownames(top_peaks))
  }
  
  g <- g + geom_point(
        data = restab[gsub("^([^A-Z].+)" , "italic('\\1')" , targetgene),],
        aes(x = logFC, y = -log10(FDR), color = "targets"),
        cex = pointsize)
  
  # add labels:
  g_labels <- unique(g_labels)
  
  g <- g + geom_label_repel(
    data = restab[g_labels,] , aes(x = logFC, y = -log10(FDR), label = gsub("-","_",rownames(restab[g_labels,]))),
    hjust = 0.1,
    size = 4, segment.alpha = 0.5, 
    segment.color = "black", 
    min.segment.length=unit(0, "cm"), parse = T) + scale_fill_manual(values=cols) + scale_color_manual(values=cols)
  g
}
```

test
```{r, eval = FALSE}
targetgene = c("acpP","fabF")
title=resname
x_limit = 5
show_sig = T
y_limit = 23
alpha=0.001
pointsize = 3
off_target_list = off_targets
pointsize = 2
minlogfc=1
title = "Volcano"
phopq = NULL
```


```{r}
# I get the links between locus tags and gene names:
pnames <- read.delim("../data/link_lt_gn.tab", header = F)
rownames(pnames) <- pnames$V2

# I also import PhoPQ related genes:
ppq_raw <- read.delim("../data/PHOPQ.tsv", header = T)
ppq <- as.character(ppq_raw$PhoPQ[ppq_raw$PhoPQ != ""])
phopqvolc <- c(pnames[pnames$V1 %in% ppq,]$V2, "PinT", "SL1344_1169", "SL1344_1168")
prefname <- ifelse(phopqvolc %in% pnames$V2 ,pnames[phopqvolc,]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
phopqvolc <- ifelse(prefname != "", prefname, phopqvolc)
phopqvolc <- c(phopqvolc, "phoP", "phoQ")
```


# Off-target analysis:
```{r}
startot <- read.delim("../data/mismatches_02_2021/pna_2mm_startregions.tab")
startot <- startot %>% filter(trans_id %in% rownames(y))

startot$genename <- sapply(startot$trans_id, function(x) ifelse(x %in% pnames$V2, pnames[x,]$V1[1] , x))

startot$mismatch_pos <- unlist(lapply(1:dim(startot)[1], function(x) {
  paste(which(strsplit(startot$target_seq, "")[[x]] != strsplit(startot$probe_seq, "")[[x]]), collapse =';')
}))

x <- "mut_6"
startot %>% filter(grepl(x, probe_id) , num_mismatch==0) %>% dim
startot %>% filter(grepl(x, probe_id) , num_mismatch==1) %>% dim
startot %>% filter(grepl(x, probe_id) , num_mismatch==2) %>% dim

```

Investigation of off-target positions:
```{r}
# get nr of bases away from the center:
startot$mismatch_pos_from_center <- sapply(startot$mismatch_pos, function(x){
  mm_pos <- abs(as.numeric(strsplit(x, ";")[[1]]) - 5.5)
  central_mm <- min(mm_pos)
  ifelse(central_mm %in% c(-Inf, Inf), NA, central_mm)
})

startot$logcpm <- unlist(sapply(1:dim(startot)[1], function(x){
  pname <- gsub(".*\\_(acpP(_PNA_mut_\\d)?)", "\\1", startot$probe_id[x])
  rname <- paste0(pname, "_vs_scr", sep="")
  lt <- startot$trans_id[x]
  x <- res_acpP_mutants_vs_scr[[rname]]$table[lt,]$logCPM
  ifelse(is.null(x), NA, x)
}))

startot$logfc <- unlist(sapply(1:dim(startot)[1], function(x){
  pname <- gsub(".*\\_(acpP(_PNA_mut_\\d)?)", "\\1", startot$probe_id[x])
  rname <- paste0(pname, "_vs_scr", sep="")
  lt <- startot$trans_id[x]
  x <- res_acpP_mutants_vs_scr[[rname]]$table[lt,]$logFC
  ifelse(is.null(x), NA, x)
}))

startot %>% filter(grepl(".*\\_(acpP(_PNA_mut_\\d)?)", probe_id) & num_mismatch == 1) %>% ggplot() + 
    geom_boxplot(aes(x=factor(mismatch_pos_from_center), y=logfc))+ geom_point(aes(x=factor(mismatch_pos_from_center), y=logfc))+
  theme_classic() + scale_size(breaks = c(0:4), range = c(1,15))
```


Now I adjust p-values (FDR), create volcano plots, histograms for the results (and save volcano plots as pdfs): 
```{r}
list_ot_0 <- list()
list_ot_1 <- list()
list_ot_2 <- list()

for (resname in names(res_acpP_mutants)){
  # adjust p-values FDR
  res_acpP_mutants[[resname]]$table$FDR <- p.adjust(res_acpP_mutants[[resname]]$table$PValue, method = "fdr")
  restab <- res_acpP_mutants[[resname]]$table
  restab$locus_tag <- rownames(restab)
  
  #add genenames (not locustags)
  prefname <- ifelse(rownames(restab) %in% pnames$V2 ,pnames[rownames(restab),]$V1, "" )
  prefname <- ifelse(isUnique(prefname), prefname, "")
  rownames(restab) <- ifelse(prefname != "", prefname, rownames(restab))
  
  restab$genes <- rownames(restab)
  
  hist(restab$PValue, breaks=100, main=resname)
  
   # check for off-targets (with 0 mismatches):
  targetgene <- gsub("_vs_ctrl", "" , resname)
  offt_zero_mm <- startot %>% filter(grepl(paste(targetgene, "$", sep = ""), probe_id) & num_mismatch==0) %>% 
    select(genename) %>% unlist
  list_ot_0[[targetgene]] <- offt_zero_mm
  
  offt_one_mm <- startot %>% filter(grepl(paste(targetgene, "$", sep = ""), probe_id) & num_mismatch==1) %>% 
    select(genename) %>% unlist
  list_ot_1[[targetgene]] <- offt_one_mm
  
  offt_two_mm <- startot %>% filter(grepl(paste(targetgene, "$", sep = ""), probe_id) & num_mismatch==2) %>% 
    select(genename) %>% unlist
  list_ot_2[[targetgene]] <- offt_two_mm
  
  off_targets <- list(zero = offt_zero_mm, one = offt_one_mm, two = offt_two_mm)
  
  #tgene_lt <- gsub(".*(ECP_.+)", "\\1", startot[grepl(targetgene, startot$probe_id),4][1])
  
  # make volcanos:
  svg(paste0("../analysis/volcanoplots_bc/",resname, ".svg"))
  print(do_volcano(restab, targetgene = c("acpP","fabF"), title=resname, 
                   x_limit = 5, show_sig = T, phopq = phopqvolc,
                   y_limit = 25,
                   alpha=0.001, pointsize = 3,
                   off_target_list = off_targets))
  dev.off()
  pdf(paste0("../analysis/volcanoplots_bc/",resname, ".pdf"))
  print(do_volcano(restab, targetgene = c("acpP","fabF"), title=resname, 
                 x_limit = 5, show_sig = T, phopq = phopqvolc,
                 y_limit = 25,
                 alpha=0.001, pointsize = 3,
                 off_target_list = off_targets))
  dev.off()
  
  # check offt with mm in outer reg:
     # check for off-targets (with 0 mismatches):
  targetgene <- gsub("_vs_ctrl", "" , resname)
  offt_zero_mm <- startot %>% filter(grepl(paste(targetgene, "$", sep = ""), probe_id) & num_mismatch==0) %>% 
    select(genename) %>% unlist
  list_ot_0[[targetgene]] <- offt_zero_mm
  
  offt_one_mm <- startot %>% filter(grepl(paste(targetgene, "$", sep = ""), probe_id) & num_mismatch==1 & 
                                      mismatch_pos_from_center > 2) %>% 
    select(genename) %>% unlist
  list_ot_1[[targetgene]] <- offt_one_mm
  
  offt_two_mm <- startot %>% filter(grepl(paste(targetgene, "$", sep = ""), probe_id) & num_mismatch==2 & 
                                      mismatch_pos_from_center > 2) %>% 
    select(genename) %>% unlist
  list_ot_2[[targetgene]] <- offt_two_mm
  
  off_targets <- list(zero = offt_zero_mm, one = offt_one_mm, two = offt_two_mm)
  
  #tgene_lt <- gsub(".*(ECP_.+)", "\\1", startot[grepl(targetgene, startot$probe_id),4][1])
  
  # make volcanos:
  svg(paste0("../analysis/volcanoplots_bc/",resname, "_ends.svg"))
  print(do_volcano(restab, targetgene = c("acpP","fabF"), title=resname, 
                   x_limit = 5, show_sig = T, phopq = phopqvolc,
                   y_limit = 25,
                   alpha=0.001, pointsize = 3,
                   off_target_list = off_targets))
  dev.off()
  pdf(paste0("../analysis/volcanoplots_bc/",resname, "_ends.pdf"))
  print(do_volcano(restab, targetgene = c("acpP","fabF"), title=resname, 
                   x_limit = 5, show_sig = T, phopq = phopqvolc,
                   y_limit = 25,
                   alpha=0.001, pointsize = 3,
                   off_target_list = off_targets))
  dev.off()
  
  #save result_table:
  dataname <- paste("../analysis/diff_exp_rawdata/", resname, ".csv", sep = "")
  write.csv(restab[order(restab$FDR),], dataname)
  #pval distributions:
  hist(restab$PValue, breaks=100, main=resname)
}


```
check for overlapping genes:
```{r}
DEgenes <- sapply(res_acpP_mutants, function (x) rownames(x$table[x$table$FDR < 0.001 & abs(x$table$logFC),]))
Reduce(intersect, DEgenes)
```




Check acpP downregulation in different mutants:
```{r}
mutants_logchange <- sapply(res_acpP_mutants, function(x) x$table["SL1344_1133",]$logFC)
mutants_pvals <- sapply(res_acpP_mutants, function(x) x$table["SL1344_1133",]$FDR)

gdata <- data.frame(mutants_logchange, sign = ifelse(mutants_pvals < 0.01 & abs(mutants_logchange) > 1, "*", ""))
gdata$names <- gsub("_vs_ctrl", "" , rownames(gdata))

gdata$names <- factor(gdata$names, levels=c("acpP_scrambled", "acpP", "acpP_PNA_mut_1", 
                                            "acpP_PNA_mut_2", "acpP_PNA_mut_3","acpP_PNA_mut_4",
                                            "acpP_PNA_mut_5","acpP_PNA_mut_6","acpP_PNA_mut_7",
                                            "acpP_PNA_mut_8","acpP_PNA_mut_9"))
gdata$names <- gdata$names
gdata$Gene <- "acpP"
g_mismatches <- gdata %>% ggplot(aes(x=names, y=-mutants_logchange)) + geom_bar(stat='identity', fill="steelblue") +
  geom_text(aes(label=sign), vjust=-0.3, size=5)+ theme_classic()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 9))+
  scale_y_continuous(limits = c(-0.1,2.5)) + xlab("") + ylab("-log2 fold change") + 
  geom_hline(yintercept=0, color="lightgrey") + 
  ggtitle(label = expression(paste("Mismatch PNAs' effect on ",italic("acpP"))))

svg("../analysis/mismatch_plot.svg")
print(g_mismatches)
dev.off()
g_mismatches
```

Same for fabF:
```{r}
mutants_logchange <- sapply(res_acpP_mutants, function(x) x$table["SL1344_1134",]$logFC)
mutants_pvals <- sapply(res_acpP_mutants, function(x) x$table["SL1344_1134",]$FDR)

gdata_fabf <- data.frame(mutants_logchange, sign = ifelse(mutants_pvals < 0.01 & abs(mutants_logchange) > 1, "*", ""))
gdata_fabf$names <- gsub("_vs_ctrl", "" , rownames(gdata_fabf))

gdata_fabf$names <- factor(gdata_fabf$names, levels=c("acpP_scrambled", "acpP", "acpP_PNA_mut_1", 
                                            "acpP_PNA_mut_2", "acpP_PNA_mut_3","acpP_PNA_mut_4",
                                            "acpP_PNA_mut_5","acpP_PNA_mut_6","acpP_PNA_mut_7",
                                            "acpP_PNA_mut_8","acpP_PNA_mut_9"))
gdata_fabf$names <- gdata_fabf$names
gdata_fabf$Gene <- "fabF"

g_mismatches <- gdata_fabf %>% ggplot(aes(x=names, y=-mutants_logchange)) + geom_bar(stat='identity', fill="steelblue") +
  geom_text(aes(label=sign), vjust=-0.3, size=5)+ theme_classic()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 9))+
  scale_y_continuous(limits = c(-0.1,2.5)) + xlab("") + ylab("-log2 fold change") + 
  geom_hline(yintercept=0, color="lightgrey") + 
  ggtitle(label = expression(paste("Mismatch PNAs' effect on ",italic("fabF"))))

svg("../analysis/mismatch_plot_fabF.svg")
print(g_mismatches)
dev.off()

gd_all <- rbind(gdata, gdata_fabf)

g_mismatches_acpp_fabf <- gd_all %>% ggplot(aes(x=names, y=-mutants_logchange, fill=Gene)) + 
  geom_bar(stat='identity', width = 0.8, position = "dodge", color="black") +
  geom_text(stat = "identity" , aes(label=sign), vjust=-0.3,hjust=0.3 , size=5,position = position_dodge(width=.8))+ 
  theme_classic() + ylab("log2 fold downregulation") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1,vjust=1, size = 9),
        axis.text.y = element_text(size=10),
        axis.title = element_text(size=14),
        title = element_text(size=15),
        legend.title = element_text(size=12),
        legend.text = element_text(face="italic", size = 12))+
  scale_y_continuous(limits = c(-0.1,2.5)) + xlab("") + 
  geom_hline(yintercept=0, color="black") + scale_fill_manual(values=c('steelblue','lightblue'))+
  ggtitle(label = expression(paste("Effect of 2-bp mismatch PNAs on ",italic("acpP"), " and ", italic("fabF"), " expression")))

svg("../analysis/mismatch_plot_acpP_fabF.svg", width = 10)
print(g_mismatches_acpp_fabf)
dev.off()
g_mismatches_acpp_fabf
```



## heatmaps rnaseq:
```{r}
mutants_logchange <- data.frame(sapply(res_acpP_mutants, function(x) x$table$logFC), 
                                row.names = rownames(res_acpP_mutants$acpP_vs_ctrl$table), 
                                genenames = prefname)
mutants_pvals <- data.frame(sapply(res_acpP_mutants, function(x) x$table$FDR), 
                            row.names = rownames(res_acpP_mutants$acpP_vs_ctrl$table),
                            genenames = prefname)

mutants_logchange <- mutants_logchange[order(mutants_pvals$acpP_vs_ctrl),]
mutants_pvals <- mutants_pvals[order(mutants_pvals$acpP_vs_ctrl),]

topdegenes <- c()
for (i in names(mutants_logchange[-12])) {
  degenes <- mutants_pvals[rownames(mutants_logchange)[abs(mutants_logchange[[i]])>1 & mutants_pvals[[i]] < 0.001],]
  degenes <- rownames(degenes[order(degenes[[i]]),])
  topdegenes <- unique(append(topdegenes, degenes[1:10]))
}
topdegenes <- topdegenes[!is.na(topdegenes)]

mutants_logchange <- mutants_logchange[topdegenes,]
mutants_pvals <- mutants_pvals[topdegenes,]

rownames(mutants_logchange) <- ifelse(mutants_logchange$genenames == "", rownames(mutants_logchange),
                                            mutants_logchange$genenames)
rownames(mutants_pvals) <- rownames(mutants_logchange)

mutants_logchange <- mutants_logchange[order(mutants_logchange$acpP_vs_ctrl, decreasing = T),][-12]
mutants_pvals <- mutants_pvals[rownames(mutants_logchange),][-12]

diff_exp <- sapply(names(mutants_pvals), function(x) {
  tf <- mutants_pvals[[x]] < 0.001 & abs(mutants_logchange[[x]])>1
  ifelse(tf, "*", " ")
  })

colnames(mutants_logchange) <- c("acpP", "scrambled", "mm1", "mm2", "mm3", "mm4", "mm5", "mm6", "mm7", "mm8", "mm9")



c1 =  circlize::colorRamp2(c(-3, 0, 3), c("blue", "white", "red"))
ht4 <- Heatmap(mutants_logchange, name = "Log2 FC",
               col = c1,
               cluster_rows = F, cluster_columns = F, show_heatmap_legend = F,
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", diff_exp[i, j]), x, y)
               }, 
               border = TRUE,
               height = unit(30, "cm"), width = unit(13, "cm"),
               #row_names_max_width = max_text_width(c(0,0),gp = gpar(fontsize = 0)),
               column_names_rot = 45)

lgd2 = Legend(col_fun = c1, title = expression("Log"[2]*" FC"), labels_gp = gpar(fontsize = 10),
             title_gp = gpar(fontsize = 15),grid_width =  unit(0.8, "cm"),
             at = c(-2, 0, 2), legend_width = unit(2, "cm"),
             labels = c("-3", "  0", "  3"), legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")

svg("../analysis/heatmap_rnaseq.svg", height = 15, width = 15)
print(ht4)
draw(lgd2, x = unit(8, "cm"), y = unit(19, "cm"), just = c("left", "bottom"))
dev.off()
```



## KEGG:
I perform the KEGG-analysis using the FRY gene set analysis tool from limma. I start with getting KEGGREST:
```{r}
library(KEGGREST)
# get link and list to get kegg info:
link_kegg <- keggLink("pathway", "sey")
list_kegg <- keggList("pathway", "sey")

kegg_pw_ids <- names(list_kegg)

#rename genes, remove ones which arent in our data:
names(link_kegg) <- gsub("sey:(.*)", "\\1", names(link_kegg)) #rename genes as locus tags
```


```{r}
link_kegg <- link_kegg[names(link_kegg) %in% c(rownames(res_acpP_mutants$acpP_vs_ctrl$table))] #remove genes not in data


idx_kegg <- sapply(kegg_pw_ids, function(x){
  x <- unique(names(link_kegg[link_kegg == x])) # choose all genes, except duplucates
})
# add phopq pw to kegg
ppq_raw <- read.delim("../data/PHOPQ.tsv", header = T)

for (c in colnames(ppq_raw)) {
  gs <- ppq_raw[[c]][ppq_raw[[c]]!=""]
  gs_lt <- pnames[pnames$V1 %in% gs,]$V2
  idx_kegg[[c]] <- gs_lt[gs_lt %in% rownames(y$counts)]
}
```

```{r}
l <- length(colnames(con_acpP_mutants))
acpp_kegg_fry <- lapply(1:l, function(x) fry(y,idx_kegg, design, con_acpP_mutants[,x]))
names(acpp_kegg_fry) <- colnames(con_acpP_mutants)
```

add KEGG terms:
```{r}
for (fryres in names(acpp_kegg_fry)) {
  acpp_kegg_fry[[fryres]][["TERM"]] <- ifelse(grepl("path",rownames(acpp_kegg_fry[[fryres]])),
                                              list_kegg[rownames(acpp_kegg_fry[[fryres]])],
                                              rownames(acpp_kegg_fry[[fryres]]))
  acpp_kegg_fry[[fryres]][["TERM"]] <- gsub("(.*) - Salmonella enterica subsp. enterica serovar Typhimurium SL1344",
                                            "\\1", acpp_kegg_fry[[fryres]][["TERM"]])
  write.csv(acpp_kegg_fry[[fryres]], paste("../analysis/pathway_analysis/", fryres, ".csv", sep = ""))
}


acpp_kegg_frysig <- lapply(acpp_kegg_fry, function(x) x[x[["FDR"]]<0.001 & x[["NGenes"]]>10,])
kegg_siggos <- c()


for (i in names(acpp_kegg_frysig)) {
  print(i)
  print(dim(acpp_kegg_frysig[[i]]))
  print(acpp_kegg_frysig[[i]][,c(1,2,4,7)])
  kegg_siggos <- c(kegg_siggos, rownames(acpp_kegg_frysig[[i]][1:10,]))  # can be modified
}

kegg_siggos <- unique(kegg_siggos[!grepl("NA", kegg_siggos)])
```



# Non-essential genes:
```{r}
y <- DGEList(non_ess_gwc[,-1], group = non_ess_test, genes = non_ess_gwc[,1,drop=FALSE])
options(digits = 3)
head(y$samples)
```


# Filtering
Now I want to filter out Genes which have very low counts across all libraries. 
I do this by creating a cutoff $$\frac {10} {L} $$
where L is the minimum library size in millions. We delete genes that are below the cutoff in at least 2 libraries:
```{r}
L <- min(y$samples$lib.size) / 1000000
cutoff <- 10/L
keep <- rowSums(cpm(y) > cutoff) >= 5
table(keep)
```
I retain only the unfiltered genes,and delete 519 genes below the threshold:
```{r}
y <- y[keep, , keep.lib.sizes=FALSE]
```



# Design matrix
I create a design matrix for the samples:
```{r}
design <- model.matrix(~0+non_ess_test)
colnames(design) <- c(levels(non_ess_test))
rownames(design) <- colnames(y$counts)
design[1:5,]
```


# Normalization

I check how the standard TMM normalization of edgeR performs. I start with calculating normalization factors:
```{r, error=FALSE}
y <- calcNormFactors(y)
y <- estimateDisp(y, design, robust = T)
```

And now I create PCA and RLE plots:
```{r}
library(RColorBrewer)
nb.cols <- 18
mycolors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)

plotPCA(cpm(y), col=mycolors[non_ess_test])
plotRLE(cpm(y), outline=FALSE, ylim=c(-1, 1), col=mycolors[non_ess_test],
        main="RLE", las=2)

# save em as svgs
svg("../analysis/PCA_non_ess_test.svg")
plotPCA(cpm(y), col=mycolors[non_ess_test])
dev.off()
svg("../analysis/RLE_non_ess_test.svg")
plotRLE(cpm(y), outline=FALSE, ylim=c(-1, 1), col=mycolors[non_ess_test],
        main="RLE", las=2)
dev.off()
```

# Differential Expression analysis using TMM
Next, I perform differential expression analysis using TMM-normalized dataset:
```{r}
con_mismatch_pnas <- makeContrasts(KFF_ackA_vs_ctrl = KFF_ackA - control,
                     KFF_ddlB_vs_ctrl = KFF_ddlB - control,
                     KFF_ompC_vs_ctrl = KFF_ompC - control,
                     KFF_penta_vs_ctrl = KFF_penta - control,
                     KFF_rpl32_vs_ctrl = KFF_rpl32 - control,
                     levels = design)

fit <- glmQLFit(y, design, robust = TRUE)


res_mismatch_pnas <- list(KFF_ackA_vs_ctrl = glmQLFTest(fit, contrast = con_mismatch_pnas[,1]),
                 KFF_ddlB_vs_ctrl = glmQLFTest(fit, contrast = con_mismatch_pnas[,2]),
                 KFF_ompC_vs_ctrl = glmQLFTest(fit, contrast = con_mismatch_pnas[,3]), 
                 KFF_penta_vs_ctrl = glmQLFTest(fit, contrast = con_mismatch_pnas[,4]), 
                 KFF_rpl32_vs_ctrl = glmQLFTest(fit, contrast = con_mismatch_pnas[,5]))
```





We now create MD, BCV and QLDisp plots to access qualiy of data:
```{r}
plotMD(y, main = "MD-plot")
abline(h=0, col="red", lty=2, lwd=2)
plotBCV(y)
plotQLDisp(fit)
```


Now I adjust p-values (FDR), create volcano plots, histograms for the results (and save volcano plots as pdfs): 
```{r}
list_ot_0 <- list()
list_ot_1 <- list()
list_ot_2 <- list()

for (resname in names(res_mismatch_pnas)){
  # adjust p-values FDR
  res_mismatch_pnas[[resname]]$table$FDR <- p.adjust(res_mismatch_pnas[[resname]]$table$PValue, method = "fdr")
  restab <- res_mismatch_pnas[[resname]]$table
  
  #add genenames (not locustags)
  prefname <- ifelse(rownames(restab) %in% pnames$V2 ,pnames[rownames(restab),]$V1, "" )
  prefname <- ifelse(isUnique(prefname), prefname, "")
  rownames(restab) <- ifelse(prefname != "", prefname, rownames(restab))
  
  restab$genes <- rownames(restab)
  
  hist(restab$PValue, breaks=100, main=resname)
  
   # check for off-targets (with 0 mismatches):
  targetgene <- gsub("_vs_ctrl", "" , resname)
  offt_zero_mm <- startot %>% filter(grepl(targetgene, probe_id) & num_mismatch==0) %>% 
    select(genename) %>% unlist
  list_ot_0[[targetgene]] <- offt_zero_mm
  
  offt_one_mm <- startot %>% filter(grepl(targetgene, probe_id) & num_mismatch==1) %>% 
    select(genename) %>% unlist
  list_ot_1[[targetgene]] <- offt_one_mm
  
  offt_two_mm <- startot %>% filter(grepl(targetgene, probe_id) & num_mismatch==2) %>% 
    select(genename) %>% unlist
  list_ot_2[[targetgene]] <- offt_two_mm
  
  off_targets <- list(zero = offt_zero_mm, one = offt_one_mm, two = offt_two_mm)
  
  #tgene_lt <- gsub(".*(ECP_.+)", "\\1", startot[grepl(targetgene, startot$probe_id),4][1])
  
  # make volcanos:
  
  svg(paste0("../analysis/volcanoplots_bc/",resname, ".svg"))
  print(do_volcano(restab, NULL, title=resname, show_sig = T,
                   x_limit = 6, phopq = phopqvolc,
                   y_limit = 23,
                   alpha=0.001, pointsize = 3,
                   off_target_list = off_targets))
  dev.off()
  
  pdf(paste0("../analysis/volcanoplots_bc/",resname, ".pdf"))
  print(do_volcano(restab, NULL, title=resname, show_sig = T,
                   x_limit = 6, phopq = phopqvolc,
                   y_limit = 23,
                   alpha=0.001, pointsize = 3,
                   off_target_list = off_targets))
  dev.off()
  
  # check offt with mm in outer reg:
     # check for off-targets (with 0 mismatches):
  targetgene <- gsub("_vs_ctrl", "" , resname)
  offt_zero_mm <- startot %>% filter(grepl(paste(targetgene, "$", sep = ""), probe_id) & num_mismatch==0) %>% 
    select(genename) %>% unlist
  list_ot_0[[targetgene]] <- offt_zero_mm
  
  offt_one_mm <- startot %>% filter(grepl(paste(targetgene, "$", sep = ""), probe_id) & num_mismatch==1 & 
                                      mismatch_pos_from_center > 3) %>% 
    select(genename) %>% unlist
  list_ot_1[[targetgene]] <- offt_one_mm
  
  offt_two_mm <- startot %>% filter(grepl(paste(targetgene, "$", sep = ""), probe_id) & num_mismatch==2 & 
                                      mismatch_pos_from_center > 3) %>% 
    select(genename) %>% unlist
  list_ot_2[[targetgene]] <- offt_two_mm
  
  off_targets <- list(zero = offt_zero_mm, one = offt_one_mm, two = offt_two_mm)
  
  #tgene_lt <- gsub(".*(ECP_.+)", "\\1", startot[grepl(targetgene, startot$probe_id),4][1])
  
  # make volcanos:
  svg(paste0("../analysis/volcanoplots_bc/",resname, "_ends.svg"))
  print(do_volcano(restab, targetgene = c("acpP","fabF"), title=resname, 
                   x_limit = 6, show_sig = T, phopq = phopqvolc,
                   y_limit = 23,
                   alpha=0.001, pointsize = 3,
                   off_target_list = off_targets))
  dev.off()
  pdf(paste0("../analysis/volcanoplots_bc/",resname, "_ends.pdf"))
  print(do_volcano(restab, targetgene = c("acpP","fabF"), title=resname, 
                   x_limit = 6, show_sig = T, phopq = phopqvolc,
                   y_limit = 23,
                   alpha=0.001, pointsize = 3,
                   off_target_list = off_targets))
  dev.off()
  
  #pval distributions:
  hist(restab$PValue, breaks=100, main=resname)
  #save result_table:
  dataname <- paste("../analysis/diff_exp_rawdata/", resname, ".csv", sep = "")
  write.csv(restab[order(restab$FDR),], dataname)
}

```

check overlapping genes:
```{r}
DEgenes <- sapply(res_mismatch_pnas, function (x) rownames(x$table[x$table$FDR < 0.01 & abs(x$table$logFC)>1,]))
overlapping <- Reduce(intersect, DEgenes)

prefname <- ifelse(overlapping %in% pnames$V2 ,pnames[overlapping,]$V1, "" )
#prefname <- ifelse(isUnique(prefname), prefname, "")
overlapping <- ifelse(prefname != "", prefname,overlapping)

```



# KEGG pathway analysis

```{r}
l <- length(colnames(con_mismatch_pnas))
list_kegg_fry <- lapply(1:l, function(x) fry(y,idx_kegg, design, con_mismatch_pnas[,x]))
names(list_kegg_fry) <- colnames(con_mismatch_pnas)
```

add KEGG terms:
```{r}
for (fryres in names(list_kegg_fry)) {
  list_kegg_fry[[fryres]][["TERM"]] <- ifelse(grepl("path",rownames(list_kegg_fry[[fryres]])),
                                              list_kegg[rownames(list_kegg_fry[[fryres]])],
                                              rownames(list_kegg_fry[[fryres]]))
  list_kegg_fry[[fryres]][["TERM"]] <- gsub("(.*) - Salmonella enterica subsp. enterica serovar Typhimurium SL1344",
                                            "\\1", list_kegg_fry[[fryres]][["TERM"]])
  write.csv(list_kegg_fry[[fryres]], paste("../analysis/pathway_analysis/", fryres, ".csv", sep = ""))
}


kegg_frysig <- lapply(list_kegg_fry, function(x) x[x[["FDR"]]<0.001 & x[["NGenes"]]>10,])


for (i in names(kegg_frysig)) {
  print(i)
  print(dim(kegg_frysig[[i]]))
  print(kegg_frysig[[i]][,c(1,2,4,7)])
  kegg_siggos <- c(kegg_siggos, rownames(kegg_frysig[[i]][1:10,]))  # can be modified
}

kegg_siggos <- unique(kegg_siggos[!grepl("NA", kegg_siggos)])


```

Create a heatmap-df  for KEGG:
```{r}
idx_kegg_char <- lapply(idx_kegg, as.character)

# I create a dataframe with mean logFC values for each significant GO-term:
mm_hm_kegg <- t(as.data.frame(lapply(idx_kegg_char[kegg_siggos], function(x){
  sapply(names(res_mismatch_pnas), function(y){
    mean(res_mismatch_pnas[[y]]$table[x,]$logFC)
  })
})))

acpp_hm_kegg <- t(as.data.frame(lapply(idx_kegg_char[kegg_siggos], function(x){
  sapply(names(res_acpP_mutants), function(y){
    mean(res_acpP_mutants[[y]]$table[x,]$logFC[!is.na(res_acpP_mutants[[y]]$table[x,]$logFC)])
  })
})))

colnames(acpp_hm_kegg) <- gsub("(.*)_vs_ctrl", "\\1", colnames(acpp_hm_kegg))
colnames(mm_hm_kegg) <- gsub("KFF_(.*)_vs_ctrl", "\\1", colnames(mm_hm_kegg))

acpp_hm_kegg <- as.data.frame(acpp_hm_kegg)
mm_hm_kegg <- as.data.frame(mm_hm_kegg)

rownames(mm_hm_kegg) <- gsub("\\.", "\\:", rownames(mm_hm_kegg))
rownames(acpp_hm_kegg) <- gsub("\\.", "\\:", rownames(acpp_hm_kegg))
```
get size of kegg terms:
```{r}

```

make heatmap:
```{r}
mm_hm_kegg <- mm_hm_kegg[order(mm_hm_kegg[,1], decreasing = T),]
acpp_hm_kegg <- acpp_hm_kegg[rownames(mm_hm_kegg),]
logchange_kegg <- cbind(acpp_hm_kegg, mm_hm_kegg)

kegg_sizes <- sapply(idx_kegg_char[rownames(logchange_kegg)], function(x) length(x))

mm_pvals <- data.frame(sapply(names(list_kegg_fry), function(x) list_kegg_fry[[x]][rownames(mm_hm_kegg),"FDR"]), 
                    row.names = rownames(mm_hm_kegg))
acpp_pvals <-data.frame(sapply(names(acpp_kegg_fry), function(x) acpp_kegg_fry[[x]][rownames(acpp_hm_kegg),"FDR"]), 
                    row.names = rownames(acpp_hm_kegg))

pvals_kegg <- cbind(acpp_pvals, mm_pvals)
#select only significant ones:
pvals_kegg <-sapply(pvals_kegg, function(x) ifelse(x<0.001, x <- "*", x<-"") )

keggpws <- list_kegg_fry$KFF_ackA_vs_ctrl[rownames(mm_hm_kegg),] [["TERM"]]


rownames(logchange_kegg) <- ifelse(!is.na(keggpws),keggpws, rownames(logchange_kegg) )
```

plot hm (save as pdf):
```{r}
col_fun = colorRamp2(c(-2,0, 2), c("blue", "white", "red"))
exp <- factor(c(rep("acpP", 11), rep("mismatch", 5)))

ht_vert <- Heatmap(logchange_kegg, cluster_rows = F, cluster_columns = F,
               name = "GO-analysis", col = col_fun,
               show_heatmap_legend = F, 
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals_kegg[i, j]), x, y)
               }, column_split = exp,
               column_names_gp = gpar(fontsize = 10),
               row_names_gp = gpar(fontsize = 10),
               row_title = NULL,
               width = unit(15, "cm"), height = unit(15, "cm"),
               column_names_rot = 45,
               right_annotation = rowAnnotation(genes = anno_barplot(kegg_sizes)))

ht_vert

lgd = Legend(col_fun = col_fun, title = expression("mean log"[2]*" FC"), #direction = "horizontal",
             title_gp = gpar(fontsize = 12), labels = c("-2", " 0"," 2"), legend_height = unit(3, "cm"),
             at = c(-2, 0, 2), border = "black",
             title_position = "leftcenter-rot")
draw(lgd)

svg("../analysis/pathway_analysis/hm_KEGG.svg", width = unit(12, "cm"),  height = unit(10, "cm"))
draw(ht_vert)
draw(lgd, x = unit(2, "cm"), y = unit(12, "cm"), just = c("left", "bottom"))
dev.off()
```

check mismatch PNAs for all sequences:
```{r}
all_res <- append(res_acpP_mutants, res_mismatch_pnas)

startot$mismatch_pos_from_center <- sapply(startot$mismatch_pos, function(x){
  mm_pos <- abs(as.numeric(strsplit(x, ";")[[1]]) - 5.5)
  central_mm <- min(mm_pos)
  ifelse(central_mm %in% c(-Inf, Inf), NA, central_mm)
})

startot$logcpm <- unlist(sapply(1:dim(startot)[1], function(x){
  pname <- gsub(".*\\_(acpP(_PNA_mut_\\d)?)", "\\1", startot$probe_id[x])
  pname <- gsub(".*_(KFF_.*)", "\\1", pname)
  rname <- paste0(pname, "_vs_ctrl", sep="")
  lt <- startot$trans_id[x]
  x <- log10(exp(all_res[[rname]]$table[lt,]$logCPM))
  ifelse(is.null(x), NA, x)
}))

startot$logfc <- unlist(sapply(1:dim(startot)[1], function(x){
  pname <- gsub(".*\\_(acpP(_PNA_mut_\\d)?)", "\\1", startot$probe_id[x])
  pname <- gsub(".*_(KFF_.*)", "\\1", pname)
  rname <- paste0(pname, "_vs_ctrl", sep="")
  lt <- startot$trans_id[x]
  x <- all_res[[rname]]$table[lt,]$logFC
  ifelse(is.null(x), NA, x)
}))

startot$minlog10FDR <- unlist(sapply(1:dim(startot)[1], function(x){
  pname <- gsub(".*\\_(acpP(_PNA_mut_\\d)?)", "\\1", startot$probe_id[x])
  pname <- gsub(".*_(KFF_.*)", "\\1", pname)
  rname <- paste0(pname, "_vs_ctrl", sep="")
  lt <- startot$trans_id[x]
  x <- -log10(all_res[[rname]]$table[lt,]$FDR)
  ifelse(is.null(x), NA, x)
}))

# add RNAseE cleavage sites for salmonella:
rnaseE <- read.delim("../data/paper_suppmat/rnasse_e_1-s2.0-S1097276516307109-mmc2.csv", header = T)
startot$RNAseE_sites <- unlist(sapply(1:dim(startot)[1], function(x){
  gname <- startot$genename[x]
  lt <- paste0(startot$trans_id[x], ";", sep="")
  sum(grepl(gname, rnaseE$Gene) | grepl(lt, rnaseE$Gene))
}))
  
mrna_HL <- read.csv("../data/paper_suppmat/LNM3_all_param.csv", header = T)
rownames(mrna_HL) <- mrna_HL$ltag
startot$mrna_HL <- unlist(sapply(1:dim(startot)[1], function(x){
  gname <- startot$trans_id[x]
  ifelse(gname %in% mrna_HL$ltag, mrna_HL$beta_WT[grepl(gname, mrna_HL$ltag)], NA)
}))

#mismatch_pos
startot %>% filter(num_mismatch %in% c(1,2)) %>% 
  #filter( logcpm>3 ) %>%
  ggplot() + 
    geom_boxplot(aes(x=factor(mismatch_pos_from_center), y=logfc ))+ 
  geom_point(aes(x=factor(mismatch_pos_from_center), y=logfc))+
  theme_classic()

#RNAseE sites
startot %>% filter(num_mismatch %in% c(1,2)) %>% 
  filter( logcpm>2 ) %>%
  ggplot() + 
  geom_point(aes(x=RNAseE_sites, y=logfc))+
  theme_classic() + scale_x_continuous(limits = c(0,20))


cols = c("0"="#440154FF","1"="#3B528BFF", "2"="#21908CFF","3"="#5DC863FF", "4"="#FDE725FF", "none"="white")
startot %>% filter(num_mismatch %in% c(0,1,2)) %>% 
  filter( logcpm>4 ) %>%
  ggplot() + geom_hline(yintercept = 0, linetype="dashed")+
  geom_point(aes(x=mrna_HL, y=logfc, bg="0"), size=4)+
  geom_point(data=startot[startot$mismatch_pos_from_center==1.5,], aes(x=mrna_HL, y=logfc, bg="1"), size=4, pch=21)+
  geom_point(data=startot[startot$mismatch_pos_from_center==2.5,], aes(x=mrna_HL, y=logfc, bg="2"), size=4, pch=21)+
  geom_point(data=startot[startot$mismatch_pos_from_center==3.5,], aes(x=mrna_HL, y=logfc, bg="3"), size=4, pch=21)+
  geom_point(data=startot[startot$mismatch_pos_from_center==4.5,], aes(x=mrna_HL, y=logfc, bg="4"), size=4, pch=21)+
  geom_point(data=startot[is.na(startot$mismatch_pos_from_center),], aes(x=mrna_HL, y=logfc, bg="none"), size=4, pch=21)+
  #geom_smooth(method=lm , color="black")+ 
  scale_fill_viridis(discrete = T) + xlim(0,1.5) + ylim(-2.5,2.5) + scale_color_viridis(discrete = T)+
  theme_minimal()+ xlab("mRNA decay rate")+ ylab("log2 FC")+
  ggtitle("Expression change vs. mRNA decay")+
  theme( axis.line = element_line()) + scale_fill_manual(values=cols)

all <- startot %>% filter(num_mismatch %in% c(0,1,2)) %>% 
  filter( logcpm>2) %>%
  ggplot(aes(x=mrna_HL, y=logfc, bg=factor("0.5"))) + 
  scale_fill_viridis(discrete = T, name= "")+
  geom_point(pch=21, size=4) + 
  ylim(-3,3) + xlim(0,1.6) +
  theme_minimal()+ xlab("mRNA decay rate")+ ylab("log2 FC")+
  ggtitle("Expression change vs. mRNA decay")+
  theme( axis.line = element_line(),
         axis.title = element_text(size = 14),
         axis.text = element_text(size = 12),
         plot.title = element_text(size = 16),
         legend.title = element_text(size = 13),
         legend.text = element_text(size = 12))

svg("../analysis/rna_decay/all_genes.svg")
print(all)
dev.off()

svg("../analysis/rna_decay/all_genes_tl.svg")
print(all+geom_smooth(method = "lm", color="black"))
dev.off()

mm_pos <- startot %>% filter(num_mismatch %in% c(0,1,2)) %>% 
  filter( logcpm>2) %>%
  ggplot(aes(x=mrna_HL, y=logfc, bg=factor(mismatch_pos_from_center))) + 
  scale_fill_viridis(discrete = T, name= "")+
  geom_point(pch=21, size=4) + 
  ylim(-3,3) + xlim(0,1.6) +
  theme_minimal()+ xlab("mRNA decay rate")+ ylab("log2 FC")+
  ggtitle("Expression change vs. mRNA decay")+
  theme( axis.line = element_line(),
         axis.title = element_text(size = 14),
         axis.text = element_text(size = 12),
         plot.title = element_text(size = 16),
         legend.title = element_text(size = 13),
         legend.text = element_text(size = 12))
svg("../analysis/rna_decay/all_genes_mm.svg")
print(mm_pos)
dev.off()



only_ends <- startot %>% filter(num_mismatch %in% c(0,1,2)) %>% 
  filter( logcpm>2 &  mismatch_pos_from_center %in% c(3.5,4.5, NA)) %>%
  ggplot(aes(x=mrna_HL, y=logfc)) + 
  scale_fill_viridis(discrete = T, name= "")+
  geom_point(aes(bg=factor(mismatch_pos_from_center)), pch=21, size=4) + 
  ylim(-3,3) + xlim(0,1.6) +
  theme_minimal()+ xlab("mRNA decay rate")+ ylab("log2 FC")+
  ggtitle("Expression change vs. mRNA decay")+
  theme( axis.line = element_line(),
         axis.title = element_text(size = 14),
         axis.text = element_text(size = 12),
         plot.title = element_text(size = 16),
         legend.title = element_text(size = 13),
         legend.text = element_text(size = 12))
svg("../analysis/rna_decay/startot.svg")
print(only_ends)
dev.off()
svg("../analysis/rna_decay/startot_tl.svg")
print(only_ends+geom_smooth(method = "lm", color = "black"))
dev.off()

```
check correlation:
```{r}
sot <- startot %>% filter(num_mismatch %in% c(0,1,2)) %>% 
    filter( logcpm>2 &  mismatch_pos_from_center %in% c(3.5,4.5, NA))
summary(lm( sot$logfc ~ sot$mrna_HL))
summary(lm( startot$logfc ~ startot$mrna_HL))
```



Packages used:
```{r}
sessionInfo()
```

